File: src/zebra/zebra.js
//------------------------------------------------------------------------------
ZebraPuzzle = function ( canvas, language ) {

    var stringsEN = [
        'The Englishman lives in the red house',
        'The Spaniard owns a dog',
        'The Japanese man is a painter',
        'The Italian drinks tea',
        'The Norwegian lives in the first house on the left',
        'The green house is on the right of the white one',
        'The photographer breeds snails',
        'The diplomat lives in the yellow house',
        'Milk is drunk in the middle house',
        'The owner of the green house drinks coffee',
        'The Norwegian\'s house is next to the blue one',
        'The violinist drinks orange juice',
        'The fox is in a house next to that of the physician',
        'The horse is in a house next to the diplomat'
    ];
    var stringsES = [
        'El ingl\xE9s vive en la casa roja',
        'El espa\xF1ol tiene un perro',
        'El japon\xE9s es un pintor',
        'El italiano toma t\xE9',
        'El noruego vive en la primera casa a la izquierda',
        'La casa verde est\xE1 a la derecha de la blanca',
        'El fot\xF3grafo cria caracoles',
        'El diplom\xE1tico vive en la casa amarilla',
        'Toman leche en la casa del medio',
        'El due\xF1o de la casa verde toma caf\xE9',
        'La casa del noruego est\xE1 al lado de la casa azul',
        'El violinista toma jugo de naranja',
        'El zorro est\xE1 en la casa al lado del doctor',
        'El caballo est\xE1 en la casa al lado del diplom\xE1tico'
    ];

    // Initial state
    var colors = ['Red', 'Yellow', 'Green', 'White', 'Blue'];
    var nations = ['Spain', 'Japan', 'Italy', 'England', 'Norway'];
    var jobs = ['Violinist', 'Physician', 'Photographer', 'Diplomat', 'Painter'];
    var pets = ['Snail', 'Fox', 'Zebra', 'Horse', 'Dog'];
    var drinks = ['Coffee', 'Milk', 'Water', 'Juice', 'Tea'];
    this.m_imageIndexes = colors.concat( nations, jobs, pets, drinks );

    // Current state
    this.m_matrix = [colors, nations, jobs, pets, drinks];

    this.m_sprites = document.getElementById( "zebra_sprites" );

    if ( !canvas || !canvas.getContext || !this.m_sprites ) {
        return;
    }
    this.m_canvas = canvas;
    this.m_context = canvas.getContext( '2d' );

    this.m_row = -1;
    this.m_col = -1;
    this.m_loaded = false;

    // Register events
    var myself = this;
    function handlerClick( event ) {
        var rect = canvas.getBoundingClientRect();
        myself.onClick( event.clientX - rect.left, event.clientY - rect.top );
    }
    canvas.addEventListener( 'click', handlerClick, false );

    // Set layout
    this.m_width = canvas.width = 350;
    canvas.height = 265 + 280;

    this.m_context.fillStyle = 'red';
    this.m_context.fillRect( 0, 0, canvas.width, canvas.height );
    this.m_context.fillStyle = 'yellow';
    this.m_context.font = '12px Verdana';
    this.m_context.fillText( 'Loading...', 150, 260 );

    var conditions = [];
    var strings = ( language === 'es' ) ? stringsES : stringsEN;
    
    // The Englishman lives in the red house
    conditions.push( new PuzzleCondition( this.m_context, strings[0], 0, 350, function () {
        return myself.m_matrix[1].indexOf( 'England' ) === myself.m_matrix[0].indexOf( 'Red' );
    } ) );
    // The Spaniard owns a dog
    conditions.push( new PuzzleCondition( this.m_context, strings[1], 1, 350, function () {
        return myself.m_matrix[1].indexOf( 'Spain' ) === myself.m_matrix[3].indexOf( 'Dog' );
    } ) );
    // The Japanese man is a painter
    conditions.push( new PuzzleCondition( this.m_context, strings[2], 2, 350, function () {
        return myself.m_matrix[1].indexOf( 'Japan' ) === myself.m_matrix[2].indexOf( 'Painter' );
    } ) );
    // The Italian drinks tea
    conditions.push( new PuzzleCondition( this.m_context, strings[3], 3, 350, function () {
        return myself.m_matrix[1].indexOf( 'Italy' ) === myself.m_matrix[4].indexOf( 'Tea' );
    } ) );
    // The Norwegian lives in the first house on the left
    conditions.push( new PuzzleCondition( this.m_context, strings[4], 4, 350, function () {
        return myself.m_matrix[1].indexOf( 'Norway' ) === 0;
    } ) );
    // The green house is on the right of the white one
    conditions.push( new PuzzleCondition( this.m_context, strings[5], 5, 350, function () {
        return myself.m_matrix[0].indexOf( 'Green' ) === myself.m_matrix[0].indexOf( 'White' ) + 1;
    } ) );
    // The photographer breeds snails
    conditions.push( new PuzzleCondition( this.m_context, strings[6], 6, 350, function () {
        return myself.m_matrix[2].indexOf( 'Photographer' ) === myself.m_matrix[3].indexOf( 'Snail' );
    } ) );
    // The diplomat lives in the yellow house
    conditions.push( new PuzzleCondition( this.m_context, strings[7], 7, 350, function () {
        return myself.m_matrix[2].indexOf( 'Diplomat' ) === myself.m_matrix[0].indexOf( 'Yellow' );
    } ) );
    // Milk is drunk in the middle house
    conditions.push( new PuzzleCondition( this.m_context, strings[8], 8, 350, function () {
        return myself.m_matrix[4].indexOf( 'Milk' ) === 2;
    } ) );
    // The owner of the green house drinks coffee
    conditions.push( new PuzzleCondition( this.m_context, strings[9], 9, 350, function () {
        return myself.m_matrix[0].indexOf( 'Green' ) === myself.m_matrix[4].indexOf( 'Coffee' )
    } ) );
    // The Norwegian’s house is next to the blue one
    conditions.push( new PuzzleCondition( this.m_context, strings[10], 10, 350, function () {
        return ( ( myself.m_matrix[1].indexOf( 'Norway' ) === myself.m_matrix[0].indexOf( 'Blue' ) + 1 )
              || ( myself.m_matrix[1].indexOf( 'Norway' ) === myself.m_matrix[0].indexOf( 'Blue' ) - 1 ) );
    } ) );
    // The violinist drinks orange juice
    conditions.push( new PuzzleCondition( this.m_context, strings[11], 11, 350, function () {
        return myself.m_matrix[2].indexOf( 'Violinist' ) === myself.m_matrix[4].indexOf( 'Juice' )
    } ) );
    // The fox is in a house next to that of the physician
    conditions.push( new PuzzleCondition( this.m_context, strings[12], 12, 350, function () {
        return ( ( myself.m_matrix[3].indexOf( 'Fox' ) === myself.m_matrix[2].indexOf( 'Physician' ) + 1 )
              || ( myself.m_matrix[3].indexOf( 'Fox' ) === myself.m_matrix[2].indexOf( 'Physician' ) - 1 ) );
    } ) );
    // The horse is in a house next to that of the diplomat
    conditions.push( new PuzzleCondition( this.m_context, strings[13], 13, 350, function () {
        return ( ( myself.m_matrix[3].indexOf( 'Horse' ) === myself.m_matrix[2].indexOf( 'Diplomat' ) + 1 )
              || ( myself.m_matrix[3].indexOf( 'Horse' ) === myself.m_matrix[2].indexOf( 'Diplomat' ) - 1 ) );
    } ) );
    this.m_conditions = conditions;
}

ZebraPuzzle.prototype.onLoaded = function () {
    this.m_loaded = true;
    this.render();
}

ZebraPuzzle.prototype.render = function () {
    var x, y, i;
    this.m_context.clearRect( 0, 0, this.m_canvas.width, this.m_canvas.height );

    // Draw initial state
    for ( x = 0; x < 5; x++ ) {
        for ( y = 0; y < 5; y++ ) {
            i = this.m_imageIndexes.indexOf( this.m_matrix[y][x] );
            this.m_context.drawImage( this.m_sprites, 70 * i, 0, 70, 53, 70 * x, 53 * y, 70, 53 );
        }
    }
    var solved = true;
    for ( i = 0; i < this.m_conditions.length; i++ ) {
        if ( !this.m_conditions[i].render() ) {
            solved = false;
        }
    }
    if ( solved ) {
        alert( "Congrats!" );
    }
};

ZebraPuzzle.prototype.onClick = function ( x, y ) {
    if ( !this.m_loaded ) {
        return;
    }

    var col = Math.floor( x / 70 );
    var row = Math.floor( y / 53 );

    if ( row === this.m_row && col === this.m_col ) {
        return;
    }
    if ( row > 4 ) {
        return;
    }

    if ( row !== this.m_row ) {
        this.m_row = row;
        this.m_col = col;
        this.render();
        this.m_context.strokeStyle = "#FF00FF";
        this.m_context.lineWidth = 3;
        this.m_context.beginPath();
        this.m_context.rect( 70 * col, 53 * row, 70, 53 );
        this.m_context.stroke();
        this.m_context.closePath();
    }
    else {
        var temp = this.m_matrix[row][col];
        this.m_matrix[row][col] = this.m_matrix[row][this.m_col];
        this.m_matrix[row][this.m_col] = temp;
        this.m_row = -1;
        this.m_col = -1;
        this.render();
    }
};

//------------------------------------------------------------------------------
PuzzleCondition = function ( context, label, index, width, evaluator ) {
    this.m_context = context;
    this.m_label = label;
    this.m_y = 265 + 20 * index;
    this.m_width = width;
    this.m_evaluator = evaluator;
}

PuzzleCondition.prototype.render = function () {
    var state = this.m_evaluator();
    this.m_context.fillStyle = state ? 'green' : 'red';
    this.m_context.fillRect( 0, this.m_y, this.m_width, 20 );
    this.m_context.fillStyle = 'yellow';
    this.m_context.font = '12px Verdana';
    this.m_context.fillText(( state ? '[\u2714] ' : '[\u2718] ' ) + this.m_label, 2, this.m_y + 14 );
    return state;
};

File: src/zebra/zebra.py
##--------------------------------------
class Generator( object ):

    def __init__( self, name, elements ):
        self.name = name
        self.elements = elements
        self.next = None
        self.generated = None
        self.fixed = None
        self.unfixed = None
        self.constrains = []
        self.current = None

    @staticmethod
    def constrain( g1, e1, g2, e2, offset = 0 ):
        g1.addConstrain( e1, g2, e2, offset )
        g2.addConstrain( e2, g1, e1, offset )   

    def addConstrain( self, e, g2, e2, offset ):
        self.constrains.append( ( g2, e2, e, offset ) )

    def chain( self, next ):
        self.next = next
        return self

    def permute( self, m, z, a = 0 ):
        if a == len( z ):
            m.append( z[:] )
        else:
            for k in range( a, len( z ) ):
                z[a], z[k] = z[k], z[a]
                self.permute( m, z, a + 1 )
                z[a], z[k] = z[k], z[a]

    def generate( self ):
        self.generated = []
        if self.fixed == None:
            self.permute( self.generated, self.elements )
        elif len( self.fixed ) > 0:
            m = []
            self.permute( m, self.unfixed )
            for t in m:
                v = self.fixed[:]
                k = 0
                for i in range( 0, len( v ) ):
                    if v[i] == None:
                        v[i] = t[k]
                        k += 1
                self.generated.append( v )
        
    def fix( self, e, p ):
        if self.fixed == None:
            self.unfixed = self.elements[:]
            self.fixed = []
            for k in self.elements:
                self.fixed.append( None )
        try:
            self.unfixed.index( e )
            if self.fixed[p] == None:
                self.fixed[p] = e
                self.unfixed.remove( e )
            else:
                self.fixed = []
        except Exception as err:
            self.fixed = []
        
    def execute( self, callback, buffer = {} ):
        constrained = False
        oldFixed = self.fixed
        oldUnfixed = self.unfixed
        if self.fixed != None: oldFixed = self.fixed[:]
        if self.unfixed != None: oldUnfixed = self.unfixed[:]

        for c in self.constrains:
            if c[0].current != None:
                constrained = True
                self.fix( c[2], c[0].current.index( c[1] ) ) 

        if self.generated == None:
            self.generate()

        for e in self.generated:
            self.current = e
            buffer[self.name] = e
            if self.next == None:
                callback( buffer )
            else:
                self.next.execute( callback, buffer )
            del buffer[self.name]
        self.current = None

        if constrained:
            self.generated = None
            self.fixed = oldFixed
            self.unfixed = oldUnfixed

##--------------------------------------
def check( z ):
    colors = z['colors']
    nations = z['nations']
    jobs = z['jobs']
    pets = z['pets']
    ## 06. The green house is on the right of the white one.
    if colors.index( 'Green' ) != colors.index( 'White' ) + 1: return
    ## 11. The Norwegian's house is next to the blue one.
    if ( nations.index( 'Norwegian' ) != colors.index( 'Blue' ) + 1 ) and ( nations.index( 'Norwegian' ) != colors.index( 'Blue' ) - 1 ): return
    ## 13. The fox is in a house next to that of the physician.
    if ( pets.index( 'Fox' ) != jobs.index( 'Physician' ) + 1 ) and ( pets.index( 'Fox' ) != jobs.index( 'Physician' ) - 1 ): return
    ## 14. The horse is in a house next to that of the diplomat.
    if ( pets.index( 'Horse' ) != jobs.index( 'Diplomat' ) + 1 ) and ( pets.index( 'Horse' ) != jobs.index( 'Diplomat' ) - 1 ): return
    print( z )

##--------------------------------------
g1 = Generator( 'colors', ['Blue', 'Green', 'Red', 'White', 'Yellow'] )
g2 = Generator( 'nations', ['Englishman', 'Italian', 'Japanese', 'Norwegian', 'Spaniard'] )
g3 = Generator( 'jobs', ['Diplomat', 'Painter', 'Photographer', 'Physician', 'Violinist'] ) 
g4 = Generator( 'pets', ['Dog', 'Fox', 'Horse', 'Snails', 'Zebra'] ) 
g5 = Generator( 'drinks', ['Coffee', 'Milk', 'Juice', 'Tea', 'Water'] ) 
# g1 = Generator( ['Yellow', 'Blue', 'Red', 'White', 'Green'] )
# g2 = Generator( ['Norwegian', 'Italian', 'Englishman', 'Spaniard', 'Japanese'] )
# g3 = Generator( ['Diplomat', 'Physician', 'Photographer', 'Violinist', 'Painter'] ) 
# g4 = Generator( ['Fox', 'Horse', 'Snails', 'Dog', 'Zebra'] ) 
# g5 = Generator( ['Water', 'Tea', 'Milk', 'Juice', 'Coffee'] ) 

## 01. The Englishman lives in the red house.
Generator.constrain( g2, 'Englishman', g1, 'Red' )
## 02. The Spaniard owns a dog.
Generator.constrain( g2, 'Spaniard', g4, 'Dog' )
## 03. The Japanese man is a painter.
Generator.constrain( g2, 'Japanese', g3, 'Painter' )
## 04. The Italian drinks tea.
Generator.constrain( g2, 'Italian', g5, 'Tea' )
## 05. The Norwegian lives in the first house on the left.
g2.fix( 'Norwegian', 0 )
## 07. The photographer breeds snails.
Generator.constrain( g3, 'Photographer', g4, 'Snails' )
## 08. The diplomat lives in the yellow house.
Generator.constrain( g3, 'Diplomat', g1, 'Yellow' )
## 09. Milk is drunk in the middle house.
g5.fix( 'Milk', 2 )
## 10. The owner of the green house drinks coffee.
Generator.constrain( g1, 'Green', g5, 'Coffee' )
## 12. The violinist drinks orange juice.
Generator.constrain( g3, 'Violinist', g5, 'Juice' )

g1.chain( g2.chain( g3.chain( g4.chain( g5 ) ) ) )
g1.execute( check )

##--------------------------------------
def check_naive( z ):
    colors = z[0]
    nations = z[1]
    jobs = z[2]
    pets = z[3]
    drinks = z[4]
    ## 01. The Englishman lives in the red house.
    if nations.index( 'Englishman' ) != colors.index( 'Red' ): return
    ## 02. The Spaniard owns a dog.
    if nations.index( 'Spaniard' ) != pets.index( 'Dog' ): return    
    ## 03. The Japanese man is a painter.
    if nations.index( 'Japanese' ) != jobs.index( 'Painter' ): return        
    ## 04. The Italian drinks tea.
    if nations.index( 'Italian' ) != drinks.index( 'Tea' ): return        
    ## 05. The Norwegian lives in the first house on the left.
    if nations.index( 'Norwegian' ) != 0: return        
    ## 06. The green house is on the right of the white one.
    if colors.index( 'Green' ) != colors.index( 'White' ) + 1: return        
    ## 07. The photographer breeds snails.
    if jobs.index( 'Photographer' ) != pets.index( 'Snails' ): return
    ## 08. The diplomat lives in the yellow house.
    if jobs.index( 'Diplomat' ) != colors.index( 'Yellow' ): return        
    ## 09. Milk is drunk in the middle house.
    if drinks.index( 'Milk' ) != 2: return        
    ## 10. The owner of the green house drinks coffee.
    if colors.index( 'Green' ) != drinks.index( 'Coffee' ): return        
    ## 11. The Norwegian's house is next to the blue one.
    if ( nations.index( 'Norwegian' ) != colors.index( 'Blue' ) + 1 ) and ( nations.index( 'Norwegian' ) != colors.index( 'Blue' ) - 1 ): return
    ## 12. The violinist drinks orange juice.
    if jobs.index( 'Violinist' ) != drinks.index( 'Juice' ): return
    ## 13. The fox is in a house next to that of the physician.
    if ( pets.index( 'Fox' ) != jobs.index( 'Physician' ) + 1 ) and ( pets.index( 'Fox' ) != jobs.index( 'Physician' ) - 1 ): return    
    ## 14. The horse is in a house next to that of the diplomat.
    if ( pets.index( 'Horse' ) != jobs.index( 'Diplomat' ) + 1 ) and ( pets.index( 'Horse' ) != jobs.index( 'Diplomat' ) - 1 ): return        
    print( z )